core/include/detray/utils/ranges/pick.hpp

```diff
@@ -298,25 +299,60 @@ class pick_view : public detray::ranges::view_interface<
 
 namespace views {
 
-template <std::input_iterator range_itr_t, std::input_iterator sequence_itr_t>
-struct pick : public pick_view<range_itr_t, sequence_itr_t> {
+template <std::input_iterator range_itr_t,
+          detray::ranges::input_range sequence_t>
+struct pick : public pick_view<range_itr_t,
+                               detray::ranges::const_iterator_t<sequence_t>> {
 
+    using sequence_itr_t = detray::ranges::const_iterator_t<sequence_t>;
     using base_type = pick_view<range_itr_t, sequence_itr_t>;
 
-    pick() = default;
+    constexpr pick() = default;
 
-    template <detray::ranges::range range_t, detray::ranges::range sequence_t>
-    DETRAY_HOST_DEVICE constexpr pick(range_t &&range, sequence_t &&seq)
+    template <detray::ranges::input_range deduced_sequence_t>
+    DETRAY_HOST_DEVICE constexpr explicit pick(deduced_sequence_t &&seq)
+        : m_seq(std::forward<deduced_sequence_t>(seq)) {}
+
+    template <detray::ranges::range range_t,
+              detray::ranges::input_range deduced_sequence_t>
+    DETRAY_HOST_DEVICE constexpr pick(range_t &&range, deduced_sequence_t &&seq)
         : base_type(std::forward<range_t>(range),
-                    std::forward<sequence_t>(seq)) {}
+                    std::forward<deduced_sequence_t>(seq)),
+          m_seq{std::forward<deduced_sequence_t>(seq)} {}
+
+    /// Call operator for range composition
+    template <detray::ranges::range range_t>
+    DETRAY_HOST_DEVICE constexpr auto operator()(range_t &&rng) {
```

Comment:
If this moves a member variable, the operator should be defined as only working on rvalues:

```suggestion
    DETRAY_HOST_DEVICE constexpr auto operator()(range_t &&rng) && {
```

Reply:
We have to be careful with naming things here, because `m_seq` is always an lvalue in this case. The question is whether or not it is a reference. If yes then indeed moving it might be dangerous; if it is not a reference then you can move it, but then you should mark the function as rvalue-only.

Reply:
Fixed